### PR TITLE
Add Phone Number exercise

### PR DESCRIPTION
**Pull request by kabiir**
- Cleaning numbers using simple regular expressions.
- Also added default parameters for phone_number.dart and encourages to explore dynamic type.
- Passes all specified tests.
- I added a new special test case which is not present in [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/phone-number/canonical-data.json) althought I'm not sure if its needed for all languages or just for my implementation.

### CODE REVIEW

**Comment by kabiir** [comment]
I believe, since the test file imports phone_number.dart and not the example.dart. Example should have a function same as to the one used in tests. It should accept a string and provide 2 types of outputs a string and null because thats what the tests are expecting. Otherwise how would the tests work? @Stargator should I still remove it from the example.dart?

File: config.json [hunk 1]
```diff
@@ -67,6 +66,19 @@
...
       "core": true,
+      "difficulty": 2,
+      "slug": "phone-number",
+      "topics": [
+        "control_flow_conditionals",
+        "control_flow_loops",
+        "dynamic",
+        "regular_expressions"
+      ],
+      "unlocked_by": "hamming",
```
Comment (Stargator) [inline config.json]: Move phone-number after bob. Set the `core` parameter to false. An! set the `unlocked_by` to `word-count`

I'm thinking this would be dependent on the Word Count exercise being merged in and then it will be the first non-core exercise to branch off.

File: config.json [hunk 2]
```diff
@@ -67,6 +66,19 @@
     },
     {
       "core": true,
+      "difficulty": 2,
+      "slug": "phone-number",
+      "topics": [
+        "control_flow_conditionals",
+        "control_flow_loops",
+        "dynamic",
```
[the remark this answers is not in the document]
Comment (kabiir) [reply config.json]: I mean dynamic return type, in dart a function can return multiple data types and we can specify that using the dynamic keyword. Similarly if we have a Map of strings that we expect to have deferent data type for each string then we would create such a map as `Map <String, dynamic>` this map can then have its keys strings only but there values can be anything.

File: exercises/phone-number/test/phone_number_test.dart [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+final phoneNumber = new PhoneNumber();
+
+void main() {
+  group("PhoneNumber: Accept only numbers - ", numbersOnlyTest);
+  group("PhoneNumber: Number length - ", numberLenghtTest);
+  group("PhoneNumber: Cleanup special characters - ", cleanUpTest);
+  group("PhoneNumber: Area/Exchange Code - ", codeRangeTest);
+  group("PhoneNumber: Edge cases - ", edgeCases);
+}
```
[the remark this answers is not in the document]
Comment (kabiir) [reply exercises/phone-number/test/phone_number_test.dart]: Did you review the edge case, should I keep it.?

File: config.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (Stargator) [reply config.json]: Understood, I would remove it as it too dart specific. The topics are suppose to be language agnostic, I think.

I haven't looked at Space Age yet, but if there's a exercise where the type returned could be different, than that would be a great way to introduce `dynamic` to users who may be new to Dart.

File: exercises/phone-number/test/phone_number_test.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (Stargator) [reply exercises/phone-number/test/phone_number_test.dart]: I think it is fine. The other test that I see as similar is "invalid with punctuation", but the difference with the edge case is that it would be valid if not for the special character, correct?

File: exercises/phone-number/lib/phone_number.dart [hunk 4]
```diff
@@ -0,0 +1,3 @@
+class PhoneNumber {
+  // Put your code here.
```
Comment (Stargator) [inline exercises/phone-number/lib/phone_number.dart]: Given the introduction to dynamic is in this. I wonder how you might feel about add a "tip" in the README or going ahead and defining the function in the class (but don't return anything. Just have a brief description of dynamic with a link to a dartlang.org page about it.

File: exercises/phone-number/test/phone_number_test.dart [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+cleanUpTest() {
+  test("cleans the number", () {
+    expect(phoneNumber.cleanNumber("(223) 456-7890"), equals("2234567890"));
+  });
```
[the remark this answers is not in the document]
Comment (kabiir) [reply exercises/phone-number/test/phone_number_test.dart]: Correct, the length of a number is 10 excluding the country code. And the edge case does have a valid number in it but because other special characters are also a part of the string it should be invalid.

File: exercises/phone-number/lib/phone_number.dart (hunk 4 above)
Comment (kabiir) [reply exercises/phone-number/lib/phone_number.dart]: That would be nice, I'll add a hint in the Readme only and keep the class empty.

**Comment by kabiir** [comment]
@Stargator, please do a final review. I think this one is also ready for merge.

**Comment by Stargator** [comment]
@kabiir What do you think of the last suggestion? It's small, but if we can determine whether or not it will be incorporated then we can get this done soon.

**Comment by kabiir** [comment]
@Stargator sorry, it's been a busy week. I'll make those changes tonight. I'll make changes to a few other prs too

File: exercises/phone-number/README.md [hunk 6]
```diff
@@ -27,5 +27,24 @@ should all produce the output
...
+String hello(int a){
+    if ( a == 0) return "a";
+    return null;
+}
+```
+To make it more clear that this function can also return `null` or more data types, use `dynamic`.
+```dart
+dynamic hello(int a){
+    if ( a == 0) return "a";
+    return null;
```
[the remark this answers is not in the document]
Comment (kabiir) [reply exercises/phone-number/README.md]: Well if we let this be. It also explains how to have one-liner `if` statements. @Stargator thoughts?

**Comment by Stargator** [comment]
@kabiir No worries, I know you are leading the effort for the binary search tree canonical-data. Thanks for all the work!

**Comment by kabiir** [comment]
Squash the commits on this, I had something weird going on.

**Comment by Stargator** [comment]
Will merge once build is complete
